keps/sig-apps/2214-indexed-job/README.md

```diff
@@ -0,0 +1,566 @@
+index.
+
+Scale down is unaffected by completion index.
+
+### Test Plan
+
+Unit, integration and E2E tests cover the following Indexed Job mechanics:
+
+  - Creation with indexed Pod names and index annotations.
+  - Scale up and down.
+  - Pod failures.
+  
+Additionally, we add unit tests for API defaulting and validation, with feature
+gate enabled and disabled.
+  
+### Graduation Criteria
+
+#### Alpha -> Beta Graduation
+
+- Complete features:
+  - Completion index in annotation
+  - Restart policy
+- Gather feedback from end users and operators' developers. Open questions:
+  - Are stable Pod names necessary?
+- Tests are in Testgrid and linked in KEP
+
+#### Beta -> GA Graduation
+
+- E2E test graduates to conformance
+- 2 examples of operators using Indexed Jobs.
+
+### Upgrade / Downgrade Strategy
+
+In the event of a kube-controller-manager upgrade, there should not be any
+existing Indexed Jobs with running Pods.
+
+In the event of a downgrade, existing Indexed Jobs will run as NonIndexed Jobs.
+The controller will track the existing Pods ignoring the completion index. New
+Pods will be created without a completion index. Existing workloads that
+expected the completion index will fail. But this is expected in a downgrade.
+
+In the event of an upgrade after a downgrade, the controller will ignore
+existing Pods without completion index for existing Indexed Jobs. This might
+cause a load spike in the cluster.
+
+If, instead of a downgrade, the feature is disabled:
+
+  - kube-apiserver sets `.spec.completionMode=NonIndexed` for new Jobs.
+  - kube-controller-manager ignores Indexed Jobs and emits a warning.
+  
+The above guarantees that the controller never creates Pods for Indexed Jobs
+without a completion index.
+
+### Version Skew Strategy
+
+This features has no node runtime implications.
+
+## Production Readiness Review Questionnaire
+
+### Feature Enablement and Rollback
+
+_This section must be completed when targeting alpha to a release._
+
+* **How can this feature be enabled / disabled in a live cluster?**
+
+  - [x] Feature gate (also fill in values in `kep.yaml`)
+    - Feature gate name: IndexedJob
+    - Components depending on the feature gate:
+      - kube-apiserver
+      - kube-controller-manager
+
+* **Does enabling the feature change any default behavior?**
+
+  No. Jobs need to opt-in with `.spec.completionMode=Indexed`.
+
+* **Can the feature be disabled once it has been enabled (i.e. can we roll back
+  the enablement)?**
+  
+  Yes. Using the feature gate is the recommended way.
+
+* **What happens if we reenable the feature if it was previously rolled back?**
+
+  The Job controller starts managing Indexed Jobs again.
+  More details covered in [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy).
+
+* **Are there any tests for feature enablement/disablement?**
+
+  Yes, unit and integration test for the feature enabled and disabled.
```

Comment:
If we're going to have something fancy for feature disablement, than we should have (at least) unit test for disablement.

You can see an example of doing that here (under review, but the idea is there):
https://github.com/kubernetes/kubernetes/pull/97058/files#diff-7826f7adbc1996a05ab52e3f5f02429e94b68ce6bce0dc534d1be636154fded3R246

Reply:
That was the intention of my text here. Reworded.